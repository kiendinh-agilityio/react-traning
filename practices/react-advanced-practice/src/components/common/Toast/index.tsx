import { useEffect, useRef } from 'react';

// Import radix ui
import { Box, Flex } from '@radix-ui/themes';

// Import for common component icon
import { CloseIcon, FailedIcon, SuccessIcon } from '@/components/common/Icons';

// Import common components
import { Button } from '@/components/common';

// Import types
import { ButtonVariant } from '@/types';

// Import styles animation for toast
import './toast.css';

interface ToastProps {
  type: string;
  message: string;
  isOpen: boolean;
  onClose: () => void;
}

const Toast = ({ type, message, isOpen, onClose }: ToastProps) => {
  const countTime = useRef<NodeJS.Timeout | null>(null);

  useEffect(() => {
    if (isOpen) {
      countTime.current = setTimeout(() => {
        onClose && onClose();
      }, 3000);
    }

    return () => {
      if (countTime.current) {
        clearTimeout(countTime.current);
      }
    };
  }, [isOpen, onClose]);

  return isOpen ? (
    <Flex
      align="center"
      justify="between"
      className={`toast ${type} w-[400px] p-4 bg-white rounded shadow-md m-auto absolute left-0 right-0 bottom-8`}
    >
      <Flex align="center" className="gap-4">
        {type === 'success' ? <SuccessIcon /> : <FailedIcon />}
        <Box as="span" className="font-regular text-base">
          {message}
        </Box>
      </Flex>
      <Button variant={ButtonVariant.Transparent} onClick={onClose}>
        <CloseIcon />
      </Button>
    </Flex>
  ) : null;
};

export default Toast;
