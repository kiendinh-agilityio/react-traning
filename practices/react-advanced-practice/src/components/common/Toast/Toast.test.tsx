import { render, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import Toast from '.';

jest.useFakeTimers();

jest.mock('@/components/common/Icons', () => ({
  CloseIcon: () => <div data-testid="close-icon">CloseIcon</div>,
  SuccessIcon: () => <div data-testid="success-icon">SuccessIcon</div>,
  FailedIcon: () => <div data-testid="failed-icon">FailedIcon</div>,
}));

describe('Toast Component', () => {
  it('renders correctly with type success', () => {
    const { container } = render(
      <Toast
        type="success"
        message="Success message"
        isOpen={true}
        onClose={jest.fn()}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with type failed', () => {
    const { container } = render(
      <Toast type="failed" message="Failed message" isOpen={true} onClose={jest.fn()} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('closes when close button is clicked', () => {
    const onCloseMock = jest.fn();

    const { getByTestId } = render(
      <Toast
        type="success"
        message="Click close test"
        isOpen={true}
        onClose={onCloseMock}
      />,
    );

    fireEvent.click(getByTestId('close-icon'));

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('calls onClose automatically after 3 seconds', () => {
    const onCloseMock = jest.fn();

    render(
      <Toast
        type="success"
        message="Auto close test"
        isOpen={true}
        onClose={onCloseMock}
      />,
    );

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('clears timeout on unmount', () => {
    const onCloseMock = jest.fn();
    const { unmount } = render(
      <Toast type="success" message="Unmount test" isOpen={true} onClose={onCloseMock} />,
    );

    unmount();

    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(onCloseMock).not.toHaveBeenCalled();
  });
});
