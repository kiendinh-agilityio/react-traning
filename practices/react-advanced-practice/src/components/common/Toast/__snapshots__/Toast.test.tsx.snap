// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Component renders correctly with type failed 1`] = `
<div>
  <div
    class="rt-Flex rt-r-ai-center rt-r-jc-space-between toast failed w-[400px] p-4 bg-white rounded shadow-md m-auto absolute left-0 right-0 bottom-8"
  >
    <div
      class="rt-Flex rt-r-ai-center gap-4"
    >
      <div
        data-testid="failed-icon"
      >
        FailedIcon
      </div>
      <span
        class="rt-Box font-regular text-base"
      >
        Failed message
      </span>
    </div>
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button  
    px-[22px] py-2.5 font-bold rounded-xl transparent cursor-pointer
   undefined"
      data-accent-color=""
    >
      <div
        data-testid="close-icon"
      >
        CloseIcon
      </div>
    </button>
  </div>
</div>
`;

exports[`Toast Component renders correctly with type success 1`] = `
<div>
  <div
    class="rt-Flex rt-r-ai-center rt-r-jc-space-between toast success w-[400px] p-4 bg-white rounded shadow-md m-auto absolute left-0 right-0 bottom-8"
  >
    <div
      class="rt-Flex rt-r-ai-center gap-4"
    >
      <div
        data-testid="success-icon"
      >
        SuccessIcon
      </div>
      <span
        class="rt-Box font-regular text-base"
      >
        Success message
      </span>
    </div>
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button  
    px-[22px] py-2.5 font-bold rounded-xl transparent cursor-pointer
   undefined"
      data-accent-color=""
    >
      <div
        data-testid="close-icon"
      >
        CloseIcon
      </div>
    </button>
  </div>
</div>
`;
